# Intro to Project

In [1]:
from csv import reader

#Open and read data to variable named hn:
hn_csv = open('HN_posts_year_to_Sep_26_2016.csv',encoding = 'UTF-8')
hn_data = reader(hn_csv)
hn = list(hn_data)
print(hn[0:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12579008', 'You have two days to comment if you want stem cells to be classified as your own', 'http://www.regulations.gov/document?D=FDA-2015-D-3719-0018', '1', '0', 'altstar', '9/26/2016 3:26'], ['12579005', 'SQLAR  the SQLite Archiver', 'https://www.sqlite.org/sqlar/doc/trunk/README.md', '1', '0', 'blacksqr', '9/26/2016 3:24'], ['12578997', 'What if we just printed a flatscreen television on the side of our boxes?', 'https://medium.com/vanmoof/our-secrets-out-f21c1f03fdc8#.ietxmez43', '1', '0', 'pavel_lishin', '9/26/2016 3:19'], ['12578989', 'algorithmic music', 'http://cacm.acm.org/magazines/2011/7/109891-algorithmic-composition/fulltext', '1', '0', 'poindontcare', '9/26/2016 3:16']]


Separate header row from the rest of the data

In [2]:
headers = hn[0]
del hn[0]

print("Headers:\n",headers)

print("\nFirst five rows:\n",hn[0:5])

Headers:
 ['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

First five rows:
 [['12579008', 'You have two days to comment if you want stem cells to be classified as your own', 'http://www.regulations.gov/document?D=FDA-2015-D-3719-0018', '1', '0', 'altstar', '9/26/2016 3:26'], ['12579005', 'SQLAR  the SQLite Archiver', 'https://www.sqlite.org/sqlar/doc/trunk/README.md', '1', '0', 'blacksqr', '9/26/2016 3:24'], ['12578997', 'What if we just printed a flatscreen television on the side of our boxes?', 'https://medium.com/vanmoof/our-secrets-out-f21c1f03fdc8#.ietxmez43', '1', '0', 'pavel_lishin', '9/26/2016 3:19'], ['12578989', 'algorithmic music', 'http://cacm.acm.org/magazines/2011/7/109891-algorithmic-composition/fulltext', '1', '0', 'poindontcare', '9/26/2016 3:16'], ['12578979', 'How the Data Vault Enables the Next-Gen Data Warehouse and Data Lake', 'https://www.talend.com/blog/2016/05/12/talend-and-Â\x93the-data-vaultÂ\x94', '1', '0', 'markgainor1', '9/26/

Separate posts beginning with Ask HN and Show HN into two different lists

In [3]:
ask_posts, show_posts, other_posts = [], [], []

for row in hn:
    title = row[1].lower()
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print ('Ask Posts:',len(ask_posts),'\nShow Posts:',len(show_posts),'\nOther Posts:',len(other_posts))

Ask Posts: 9139 
Show Posts: 10158 
Other Posts: 273822


Check if ask posts or show posts receive more comments

In [4]:
def find_num_comments(posts):
    current_list = []
    for row in posts:
        num_comments = int(row[4])
        current_list.append(num_comments)
    return current_list

def calc_avg(num_comments):
    avg = sum(num_comments)/len(num_comments)
    return avg

total_ask_comments = find_num_comments(ask_posts)
total_show_comments = find_num_comments(show_posts)
avg_ask_comments = calc_avg(total_ask_comments)
avg_show_comments = calc_avg(total_show_comments)

print ('Average number of comments for ask posts:', avg_ask_comments)
print ('Average number of comments for show posts:', avg_show_comments)

Average number of comments for ask posts: 10.393478498741656
Average number of comments for show posts: 4.886099625910612


Ask posts receive more comments (10.39 per post) than show posts (4.89 per post)

Moving forward, only the ask_posts list will be analysed since it received more comments per post. The next step is to analyse the number of posts created per hour of the day along with the total number of comments made in each hour.

In [34]:
import datetime as dt

result_list = []
for row in ask_posts:
    temp_list = [row[6],int(row[4])]
    result_list.append(temp_list)
                       
counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    date_object = dt.datetime.strptime(row[0],"%m/%d/%Y %H:%M")
    hour = date_object.hour
    num_comments = row[1]
    
    if hour in counts_by_hour:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += row[1]
    else:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = row[1]
        
print("Count of posts created by hour:\n",counts_by_hour)
print('\n')
print("Count of comments made by hour:\n",comments_by_hour)

Count of posts created by hour:
 {2: 269, 1: 282, 22: 383, 21: 518, 19: 552, 17: 587, 15: 646, 14: 513, 13: 444, 11: 312, 10: 282, 9: 222, 7: 226, 3: 271, 23: 343, 20: 510, 16: 579, 8: 257, 0: 301, 18: 614, 12: 342, 4: 243, 6: 234, 5: 209}


Count of comments made by hour:
 {2: 2996, 1: 2089, 22: 3372, 21: 4500, 19: 3954, 17: 5547, 15: 18525, 14: 4972, 13: 7245, 11: 2797, 10: 3013, 9: 1477, 7: 1585, 3: 2154, 23: 2297, 20: 4462, 16: 4466, 8: 2362, 0: 2277, 18: 4877, 12: 4234, 4: 2360, 6: 1587, 5: 1838}


Calculate avg number of comments per post for posts created during each hour of teh day:

In [41]:
avg_by_hour = []
for hour in counts_by_hour:
    avg_by_hour.append([hour,comments_by_hour[hour]/counts_by_hour[hour]])
    
print(avg_by_hour)
    

[[2, 11.137546468401487], [1, 7.407801418439717], [22, 8.804177545691905], [21, 8.687258687258687], [19, 7.163043478260869], [17, 9.449744463373083], [15, 28.676470588235293], [14, 9.692007797270955], [13, 16.31756756756757], [11, 8.96474358974359], [10, 10.684397163120567], [9, 6.653153153153153], [7, 7.013274336283186], [3, 7.948339483394834], [23, 6.696793002915452], [20, 8.749019607843136], [16, 7.713298791018998], [8, 9.190661478599221], [0, 7.5647840531561465], [18, 7.94299674267101], [12, 12.380116959064328], [4, 9.7119341563786], [6, 6.782051282051282], [5, 8.794258373205741]]


Create a new list to sort the list by highest avg number of comments per post for posts created during each hour of the day

In [65]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])
    
print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour, reverse = True)

print('\nTop 5 Hours for Ask Posts Comments')
for entry in sorted_swap[0:5]:
    avg = entry[0]
    hour = entry[1]
    print('{hour}:00: {avg:.2f} average comments per post'.format(hour = hour, avg = avg))



[[11.137546468401487, 2], [7.407801418439717, 1], [8.804177545691905, 22], [8.687258687258687, 21], [7.163043478260869, 19], [9.449744463373083, 17], [28.676470588235293, 15], [9.692007797270955, 14], [16.31756756756757, 13], [8.96474358974359, 11], [10.684397163120567, 10], [6.653153153153153, 9], [7.013274336283186, 7], [7.948339483394834, 3], [6.696793002915452, 23], [8.749019607843136, 20], [7.713298791018998, 16], [9.190661478599221, 8], [7.5647840531561465, 0], [7.94299674267101, 18], [12.380116959064328, 12], [9.7119341563786, 4], [6.782051282051282, 6], [8.794258373205741, 5]]

Top 5 Hours for Ask Posts Comments
15:00: 28.68 average comments per post
13:00: 16.32 average comments per post
12:00: 12.38 average comments per post
2:00: 11.14 average comments per post
10:00: 10.68 average comments per post
